File: server/models/Alert.js
import mongoose from 'mongoose';

const alertSchema = new mongoose.Schema({
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  type: {
    type: String,
    enum: ['close-call', 'life-threat', 'obstacle', 'warning', 'info'],
    default: 'info'
  },
  severity: {
    type: String,
    enum: ['low', 'medium', 'high', 'critical'],
    default: 'medium'
  },
  description: {
    type: String,
    required: true
  },
  imageRef: {
    type: String
  },
  detectedObjects: [{
    object: String,
    confidence: Number,
    distance: String
  }],
  location: {
    latitude: Number,
    longitude: Number
  },
  modelResponse: {
    type: String
  },
  acknowledged: {
    type: Boolean,
    default: false
  },
  acknowledgedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  acknowledgedAt: Date,
  emailsSent: [{
    recipientEmail: String,
    sentAt: Date,
    status: String
  }],
  createdAt: {
    type: Date,
    default: Date.now
  }
});

// Index for efficient queries
alertSchema.index({ userId: 1, createdAt: -1 });
alertSchema.index({ severity: 1, acknowledged: 1 });

export default mongoose.model('Alert', alertSchema);

File: backend/app/models/alert.py
"""
Drishti AI - Alert Model

MongoDB document model for alerts/warnings.
"""

from beanie import Document, Indexed
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class AlertType(str, Enum):
    """Alert type enumeration."""
    CLOSE_CALL = "close-call"
    LIFE_THREAT = "life-threat"
    OBSTACLE = "obstacle"
    WARNING = "warning"
    INFO = "info"


class AlertSeverity(str, Enum):
    """Alert severity enumeration."""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class DetectedObject(BaseModel):
    """Detected object embedded document."""
    object: str
    confidence: float = 0.0
    distance: str = "unknown"


class Location(BaseModel):
    """Location embedded document."""
    latitude: Optional[float] = None
    longitude: Optional[float] = None


class EmailSent(BaseModel):
    """Email sent record embedded document."""
    recipient_email: str
    sent_at: datetime = Field(default_factory=datetime.utcnow)
    status: str = "sent"


class Alert(Document):
    """Alert document model."""
    
    # Reference to user
    user_id: Indexed(str)  # Store as string ID
    
    # Alert details
    type: AlertType = AlertType.INFO
    severity: AlertSeverity = AlertSeverity.MEDIUM
    description: str
    
    # Image reference
    image_ref: Optional[str] = None
    
    # Detected objects
    detected_objects: List[DetectedObject] = Field(default_factory=list)
    
    # Location
    location: Optional[Location] = None
    
    # Model response
    model_response: Optional[str] = None
    
    # Acknowledgement
    acknowledged: bool = False
    acknowledged_by: Optional[str] = None  # User ID as string
    acknowledged_at: Optional[datetime] = None
    
    # Email tracking
    emails_sent: List[EmailSent] = Field(default_factory=list)
    
    # Timestamp
    created_at: datetime = Field(default_factory=datetime.utcnow)
    
    class Settings:
        name = "alerts"
        indexes = [
            [("user_id", 1), ("created_at", -1)],
            [("severity", 1), ("acknowledged", 1)]
        ]
